src/v/cloud_roles/refresh_credentials.h

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "cloud_roles/logger.h"
+#include "cloud_roles/probe.h"
+#include "cloud_roles/request_response_helpers.h"
+#include "model/metadata.h"
+#include "s3/signature.h"
+#include "seastarx.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/util/noncopyable_function.hh>
+
+#include <memory>
+
+namespace cloud_roles {
+
+struct retry_params {
+    ss::lowres_clock::duration backoff_ms;
+    uint8_t max_retries;
+};
+
+class refresh_credentials {
+public:
+    class impl {
+    public:
+        impl(
```

Comment:
is this a case where you can introduce the impl interface in a commit and then each of the implementations in separate commits?

Reply:
added them to different commits, although the base class also introduces aws impl